### PR TITLE
Fix inconsistent matrix representation

### PR DESCRIPTION
**Pull request by joannacirillo**
### Context

fix #2325.
In issue #2325, it was mentionned that it is preferable to have only one format for matrices. For performance reason, it should be number[] .

### Results

The matrix format used in the math module (and therefore wherever those functions are called) are in number[] format. Tests have been added for these functions.

### Changes

Changed : testMath, math, TextureMapToPlane, MeshClipPlane, OBBTree, LandmarkTransform, ImageMapper, Line, Triangle
- [x] Documentation and TypeScript definitions were updated to match those changes

### PR and Code Checklist

- [x] [semantic-release](https://github.com/semantic-release/semantic-release) commit messages
- [x] Run `npm run reformat` to have correctly formatted code

### Testing

- [x] This change adds or fixes unit tests 
- [x] Tested environment:
  - **vtk.js**: 24.5.2
  - **OS**: Windows 10
  - **Browser**: Chrome 100.0.4896.88

### CODE REVIEW

**Comment by joannacirillo** [comment]
@luciemac

File: Sources/Filters/Texture/TextureMapToPlane/index.js [hunk 1]
```diff
@@ -87,19 +87,23 @@ function vtkTextureMapToPlane(publicAPI, model) {
...
     }
 
-    matrix[0] = v;
-    matrix[1] = c2;
-    matrix[2] = c3;
+    for (i = 0; i < 3; i++) {
+      matrix[i] = v[i];
+      matrix[i + 3] = c2[i];
+      matrix[i + 6] = c3[i];
+    }
```
Comment (finetjul) [inline Sources/Filters/Texture/TextureMapToPlane/index.js:100]: for performance reasons, I would simply do `matrix = [v[0], v[1], v[2], c2[0], ...]` instead of a for loop.

File: Sources/Rendering/OpenGL/PolyDataMapper/test/testMoreClippingPlanes.js [hunk 2]
```diff
@@ -96,9 +96,15 @@ test('Test PolyDataMapper Clipping Planes 2', (t) => {
...
-    [0, 0, 1],
+    Math.cos(theta),
+    Math.sin(theta),
+    0,
+    -Math.sin(theta),
+    Math.cos(theta),
+    0,
+    0,
+    0,
+    1,
```
Comment (finetjul) [inline Sources/Rendering/OpenGL/PolyDataMapper/test/testMoreClippingPlanes.js:107]: please list 3 elements per line (and disable linter temporarily

File: Sources/Common/Core/Math/index.js [hunk 3]
```diff
@@ -383,7 +375,7 @@ export function gaussianWeight(mean, variance, position) {
 export function outer2D(x, y, out_2x2) {
   for (let i = 0; i < 2; i++) {
     for (let j = 0; j < 2; j++) {
-      out_2x2[i][j] = x[i] * y[j];
+      out_2x2[i * 2 + j] = x[i] * y[j];
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:378]: replace nested for loops with hardcoded exhaustive assignment (can be done in a separate PR)

File: Sources/Common/Core/Math/index.js [hunk 4]
```diff
@@ -17,14 +17,6 @@ function notImplemented(method) {
   return () => vtkErrorMacro(`vtkMath::${method} - NOT IMPLEMENTED`);
 }
 
-function vtkSwapVectors3(v1, v2) {
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:20]: you could keep it with the following signature: `vtkSwapMatrixColumns(matrix, column1, column2)` (or vtkSwapMatrixRows?)

File: Sources/Common/Core/Math/index.js [hunk 5]
```diff
@@ -530,44 +532,29 @@ export function linearSolve3x3(mat_3x3, x_3, y_3) {
...
-      a_3x3[1][1] * b_3x3[1][i] +
-      a_3x3[1][2] * b_3x3[2][i];
-    tmp[2][i] =
-      a_3x3[2][0] * b_3x3[0][i] +
-      a_3x3[2][1] * b_3x3[1][i] +
-      a_3x3[2][2] * b_3x3[2][i];
+  const copyA = createArray(9);
+  const copyB = createArray(9);
+  for (let j = 0; j < 9; j++) {
+    copyA[j] = a_3x3[j];
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:548]: it is useless to `createArray` filled with 0s (more costly than `new Array(9)` ) and overwrite it just after

Comment (finetjul) [reply Sources/Common/Core/Math/index.js:548]: use `a_3x3.slice(0, 9)`

File: Sources/Common/Core/Math/index.js [hunk 6]
```diff
@@ -577,47 +564,57 @@ export function multiplyMatrix(a, b, rowA, colA, rowB, colB, out_rowXcol) {
...
   }
 
+  const copyA = createArray(9);
+  const copyB = createArray(9);
+  for (let j = 0; j < rowA * colA; j++) {
+    copyA[j] = a[j];
+  }
+  for (let j = 0; j < rowB * colB; j++) {
+    copyB[j] = b[j];
+  }
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:574]: why a copy is needed here ? please add comment

File: Sources/Common/Core/Math/index.js [hunk 7]
```diff
@@ -577,47 +564,57 @@ export function multiplyMatrix(a, b, rowA, colA, rowB, colB, out_rowXcol) {
...
       // sum for this point
       for (let k = 0; k < colA; k++) {
-        out_rowXcol[i][j] += a[i][k] * b[k][j];
+        out_rowXcol[i * colB + j] += copyA[i * colA + k] * copyB[j + colB * k];
       }
     }
   }
 }
 
 export function transpose3x3(in_3x3, outT_3x3) {
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:589]: why not keep the same logic as the previous implementation (using a tmp value)?

File: Sources/Common/Core/Math/index.js [hunk 8]
```diff
@@ -634,38 +631,43 @@ export function invert3x3(in_3x3, outI_3x3) {
 
   // Divide by the determinant
   const det = a1 * d1 + b1 * d2 + c1 * d3;
+  if (det === 0) {
+    vtkWarningMacro('Matrix has determinant 0');
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:635]: has 0 determinant

File: Sources/Common/Core/Math/index.js [hunk 9]
```diff
@@ -634,38 +631,43 @@ export function invert3x3(in_3x3, outI_3x3) {
...
 }
 
+export function identity3x3(mat_3x3) {
+  for (let i = 0; i < 9; i++) {
+    mat_3x3[i] = 0.0;
+    if (i === 0 || i === 4 || i === 8) {
+      mat_3x3[i] = 1.0;
+    }
+  }
+}
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:669]: previous implementation was more elegant (and faster)

File: Sources/Common/Core/Math/index.js [hunk 10]
```diff
@@ -757,23 +766,28 @@ export function jacobiN(a, n, w, v) {
...
-    for (iq = 0; iq < n; iq++) {
-      v[ip][iq] = 0.0;
+  for (ip = 0; ip < n * n; ip++) {
+    v[ip] = 0.0;
+    if (ip % (n + 1) === 0) {
+      v[ip] = 1.0;
     }
-    v[ip][ip] = 1.0;
+    // v[ip] = ip % (n + 1) === 0 ? 1.0 : 0.0;
   }
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:788]: previous implementation was more elegant.

(remove comment)

you could create an identity function that takes n as parameter

File: Sources/Common/Core/Math/index.js [hunk 11]
```diff
@@ -757,23 +766,28 @@ export function jacobiN(a, n, w, v) {
   let tmp;
   const b = createArray(n);
   const z = createArray(n);
+  const a_temp = createArray(n * n);
+  for (let index = 0; index < n * n; index++) {
+    a_temp[index] = a[index];
+  }
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:772]: `a_temp = [...a]` or  `a.slice();`

File: Sources/Common/Core/Math/index.js [hunk 12]
```diff
@@ -890,66 +904,60 @@ export function jacobiN(a, n, w, v) {
...
+        v[i * n + j] *= -1.0;
       }
     }
   }
+  // for (i = 0; i < n; i++) {
+  //   w[i] = roundNumber(w[i], 6);
+  //   for (j = 0; j < n; j++) {
+  //     v[i * n + j] = roundNumber(v[i * n + j], 6);
+  //   }
+  // }
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:923]: to remove

File: Sources/Common/Core/Math/index.js (hunk 11 above)
Comment (finetjul) [reply Sources/Common/Core/Math/index.js:772]: why not `copyA` to be consistent with the rest of your changes ?

File: Sources/Common/Core/Math/index.js [hunk 13]
```diff
@@ -798,21 +812,21 @@ export function jacobiN(a, n, w, v) {
...
 
         // after 4 sweeps
         if (
           i > 3 &&
           Math.abs(w[ip]) + g === Math.abs(w[ip]) &&
           Math.abs(w[iq]) + g === Math.abs(w[iq])
         ) {
-          a[ip][iq] = 0.0;
-        } else if (Math.abs(a[ip][iq]) > tresh) {
+          a_temp[ip * n + iq] = 0.0;
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:823]: it no longer modifies input `a`, is it on purpose ?

File: Sources/Common/Core/Math/index.js [hunk 14]
```diff
@@ -1083,40 +1104,38 @@ export function diagonalize3x3(a_3x3, w_3, v_3x3) {
...
-    identity3x3(v_3x3);
+    v_3x3[0] = 1.0;
+    v_3x3[4] = 1.0;
+    v_3x3[8] = 1.0;
+    v_3x3[1] = 0.0;
+    v_3x3[2] = 0.0;
+    v_3x3[3] = 0.0;
+    v_3x3[5] = 0.0;
+    v_3x3[6] = 0.0;
+    v_3x3[7] = 0.0;
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:1120]: use identity3x3

File: Sources/Common/Core/Math/index.js [hunk 15]
```diff
@@ -1126,29 +1145,49 @@ export function diagonalize3x3(a_3x3, w_3, v_3x3) {
...
+        vectTmp1,
+        [v_3x3T[maxI * 3], v_3x3T[maxI * 3 + 1], v_3x3T[maxI * 3 + 2]],
+        vectTmp2
+      );
+      for (let t = 0; t < 3; t++) {
+        v_3x3T[k * 3 + t] = vectTmp1[t];
+      }
+      for (let t = 0; t < 3; t++) {
+        v_3x3T[j * 3 + t] = vectTmp2[t];
+      }
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:1187]: merge both loops into one loop

File: Sources/Common/Core/Math/index.js [hunk 16]
```diff
@@ -1126,29 +1145,49 @@ export function diagonalize3x3(a_3x3, w_3, v_3x3) {
...
+      v_3x3T[j * 3 + 2] = 0.0;
+      v_3x3T[j * 3 + j] = 1.0;
+      const vectTmp1 = createArray(3);
+      cross(
+        [v_3x3T[maxI * 3], v_3x3T[maxI * 3 + 1], v_3x3T[maxI * 3 + 2]],
+        [v_3x3T[j * 3], v_3x3T[j * 3 + 1], v_3x3T[j * 3 + 2]],
+        vectTmp1
+      );
+      normalize(vectTmp1);
+      const vectTmp2 = createArray(3);
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:1176]: not needed -> `const vecTmp2 = cross (...., []);`

File: Sources/Common/Core/Math/index.js [hunk 17]
```diff
@@ -1083,40 +1104,38 @@ export function diagonalize3x3(a_3x3, w_3, v_3x3) {
...
+    v_3x3[6] = 0.0;
+    v_3x3[7] = 0.0;
     return;
   }
 
   // transpose temporarily, it makes it easier to sort the eigenvectors
-  transpose3x3(v_3x3, v_3x3);
+  // copy eigenvectors vector before transpose
+  const v_3x3T = createArray(9);
+  transpose3x3(v_3x3, v_3x3T);
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:1127]: why create a new variable, simply use `v_3x3`, no ? like before

File: Sources/Common/Core/Math/test/testMath.js [hunk 18]
```diff
@@ -1,6 +1,14 @@
 import test from 'tape-catch';
 import * as vtkMath from 'vtk.js/Sources/Common/Core/Math';
 
+const createArray = (size = 3) => {
```
Comment (finetjul) [inline Sources/Common/Core/Math/test/testMath.js:4]: export it from vtkMath if you need it here instead of re-declaring it

**Review by finetjul — CHANGES_REQUESTED** [review]

File: Sources/Common/Core/Math/index.js (hunk 13 above)
Comment (joannacirillo) [reply Sources/Common/Core/Math/index.js:823]: Yes, the modification of `a` during computation of eigen values and vectors made the test of diagonalize3x3 function tricky. The verification of `D = P^-1 A P` is done by computing the eigen vectors on one hand and computing the diagonalization on the another hand. As `a` was changed in the computation of the eigen values and vectors `D = P^-1 A P` was false even though the computation of `D` and `P` was correct. Since the modified value of `a` was not meaningfull, I thought it was better to copy the matrix inside of jacobiN before doing the computation than having to copy it whenever we want to pass it to jacobiN. It is not the same implementation as vtk C++ though.

File: Sources/Common/Core/Math/index.js (hunk 7 above)
Comment (joannacirillo) [reply Sources/Common/Core/Math/index.js:590]: Do you mean avoiding the copy of a and b by storing the result in a temporary Array and then copy it into out_rowXcol ?

File: Sources/Common/Transform/LandmarkTransform/test/testLandmarkTransform.js [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+  transform.update();
+  transformMatrix = transform.getMatrix();
+  t.ok(
+    vtkMath.areEquals(
+      transformMatrix,
+      [
+        1, -4.9960036106613045e-17, -1.1657341758720898e-16, 0,
+        4.9960036106613045e-17, 1, -4.9960036106612817e-17, 0,
+        1.1657341758720898e-16, 4.9960036106612817e-17, 1, 0,
+        2.9999999999999996, 3, 3.000000000000001, 1,
```
[the remark this answers is not in the document]
Comment (joannacirillo) [reply Sources/Common/Transform/LandmarkTransform/test/testLandmarkTransform.js]: I rounded the values to 1e-6 (since it is the order used in `areEquals`)

File: Sources/Common/Core/Math/index.js [hunk 20]
```diff
@@ -881,7 +881,7 @@ export function jacobiN(a, n, w, v) {
     tmp = w[k];
     for (i = j + 1; i < n; i++) {
       // boundary incorrect, shifted already
-      if (w[i] >= tmp) {
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js]: such fix, if needed, must be done in a separate PR

File: Sources/Filters/Texture/TextureMapToPlane/index.js [hunk 21]
```diff
@@ -87,7 +87,7 @@ function vtkTextureMapToPlane(publicAPI, model) {
     const c1 = [m[0], m[1], m[2]];
```
Comment (finetjul) [inline Sources/Filters/Texture/TextureMapToPlane/index.js:87]: line 61-63 no longer seem to be needed

File: Sources/Filters/Texture/TextureMapToPlane/index.js [hunk 22]
```diff
@@ -87,7 +87,7 @@ function vtkTextureMapToPlane(publicAPI, model) {
...
     const c2 = [m[3], m[4], m[5]];
     const c3 = [m[6], m[7], m[8]];
-    const matrix = [c1, c2, c3];
+    const matrix = m;
```
Comment (finetjul) [inline Sources/Filters/Texture/TextureMapToPlane/index.js]: this new variable matrix does not seem needed anymore, you could use `m` instead

File: Sources/Common/Core/Math/index.js [hunk 23]
```diff
@@ -17,7 +17,17 @@ function notImplemented(method) {
   return () => vtkErrorMacro(`vtkMath::${method} - NOT IMPLEMENTED`);
 }
 
-function createArray(size = 3) {
+// Swap columns for n by n matrix
+function swapRowsOrColumnsMatrix_nxn(matrix, n, col1, col2) {
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js]: why name it "swapRows" if it only swap columns ? -> swapColumnsMatrix_nxn

File: Sources/Common/Core/Math/index.js [hunk 24]
```diff
@@ -660,14 +661,21 @@ export function determinant3x3(mat_3x3) {
...
-      mat_3x3[i] = 1.0;
-    }
+  for (let i = 0; i < 3; i++) {
+    /* eslint-disable-next-line no-multi-assign */
+    mat_3x3[i * 3] = mat_3x3[i * 3 + 1] = mat_3x3[i * 3 + 2] = 0;
+    mat_3x3[i * 3 + i] = 1;
   }
 }
 
+export function identity(n) {
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:672]: pass mat as function input parameter

File: Sources/Common/Core/Math/index.js [hunk 25]
```diff
@@ -766,10 +774,7 @@ export function jacobiN(a, n, w, v) {
   let tmp;
   const b = createArray(n);
   const z = createArray(n);
-  const a_temp = createArray(n * n);
-  for (let index = 0; index < n * n; index++) {
-    a_temp[index] = a[index];
-  }
+  const copyA = a.slice();
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js]: use spread operator

**Review by finetjul — CHANGES_REQUESTED** [review]

File: Sources/Common/Core/Math/index.js (hunk 23 above)
Comment (joannacirillo) [reply Sources/Common/Core/Math/index.js]: Some confusion on my side (due to the fact that both swaping columns and rows are required in different part of the code), last commit should solve this.

File: Sources/Common/Core/Math/index.js (hunk 10 above)
Comment (joannacirillo) [reply Sources/Common/Core/Math/index.js]: I think last commit solves this problem.

**Review by finetjul — CHANGES_REQUESTED** [review]

File: Sources/Common/Core/Math/test/testMath.js [hunk 26]
```diff
@@ -162,7 +162,13 @@ test('Test Identity3x3', (t) => {
...
+  // prettier-ignore
+  const m = [
+    0, 4, 0, 0,
+    2, 2, 1, 0,
+    0, 0, 9, 0,
+    0, 3, 7, 0]
+  vtkMath.identity(m, 4);
   // prettier-ignore
   t.ok(
     vtkMath.areEquals(m, [
```
[the remark this answers is not in the document]
Comment (joannacirillo) [reply Sources/Common/Core/Math/test/testMath.js:174]: Is it required that m can be any size before being set to Identity ?

[the remark this answers is not in the document]
Comment (finetjul) [reply Sources/Common/Core/Math/test/testMath.js:174]: no, just empty.
fyi, the following works:
```
const m = [];
m[10] = 0;
```

File: Sources/Filters/General/OBBTree/index.js [hunk 27]
```diff
@@ -153,39 +153,37 @@ function vtkOBBTree(publicAPI, model) {
...
-    min[2] = v[2][2];
+    max[0] = v[0];
+    max[1] = v[3];
+    max[2] = v[6];
+    mid[0] = v[1];
+    mid[1] = v[4];
+    mid[2] = v[7];
+    min[0] = v[2];
+    min[1] = v[5];
+    min[2] = v[8];
```
[the remark this answers is not in the document]
Comment (joannacirillo) [reply Sources/Filters/General/OBBTree/index.js:181]: Yes. `a` represents the covariance matrix on the three coordinate axis thus it is a 3x3 matrix. The matrix `v` is also 3x3 as it contains the 3 eigen vectors of `a`.

File: Sources/Common/Core/Math/index.js (hunk 20 above)
Comment (joannacirillo) [reply Sources/Common/Core/Math/index.js]: It has been reverted in further commits. This change does not appear on the final changes anymore.

File: Sources/Common/Transform/LandmarkTransform/index.js [hunk 28]
```diff
@@ -13,23 +13,6 @@ function vtkLandmarkTransform(publicAPI, model) {
   // Set our className
   model.classHierarchy.push('vtkLandmarkTransform');
 
-  // Convert a mat4 matrix to a Matrix 2 dimensions
-  function mat4To2DArray(mat) {
```
[the remark this answers is not in the document]
Comment (finetjul) [reply Sources/Common/Transform/LandmarkTransform/index.js:17]: it seems that you are actually doing such conversion in some parts of your code, don't you ? 
If so, it means there is a need and it would reduce code duplciation.

File: Sources/Common/Core/Math/index.js [hunk 29]
```diff
@@ -1202,22 +1209,22 @@ export function diagonalize3x3(a_3x3, w_3, v_3x3) {
 
 export function singularValueDecomposition3x3(a_3x3, u_3x3, w_3, vT_3x3) {
   let i;
-  const B = [createArray(3), createArray(3), createArray(3)];
+  const B = createArray(9);
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:1212]: should not be needed to zero initialize: `const B = []` is enough (or `const B = new Array(9)`)

File: Sources/Common/Core/Math/index.js [hunk 30]
```diff
@@ -1202,22 +1209,22 @@ export function diagonalize3x3(a_3x3, w_3, v_3x3) {
...
 
   // copy so that A can be used for U or VT without risk
   for (i = 0; i < 3; i++) {
-    B[0][i] = a_3x3[0][i];
-    B[1][i] = a_3x3[1][i];
-    B[2][i] = a_3x3[2][i];
+    B[i] = a_3x3[i];
+    B[3 + i] = a_3x3[3 + i];
+    B[6 + i] = a_3x3[6 + i];
   }
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:1219]: actually, isn't the following good enough: ? `const B = [...a_3x3]`

you might want to do a pass on all your modified code to make sure you have not let such unneeded complexity around.

File: Sources/Common/Core/Math/index.js [hunk 31]
```diff
@@ -1202,22 +1209,22 @@ export function diagonalize3x3(a_3x3, w_3, v_3x3) {
...
   // temporarily flip if determinant is negative
   const d = determinant3x3(B);
   if (d < 0) {
     for (i = 0; i < 3; i++) {
-      B[0][i] = -B[0][i];
-      B[1][i] = -B[1][i];
-      B[2][i] = -B[2][i];
+      B[i] = -B[i];
+      B[3 + i] = -B[3 + i];
+      B[6 + i] = -B[6 + i];
```
Comment (finetjul) [inline Sources/Common/Core/Math/index.js:1227]: loop on 9 instead of 3

File: Sources/Common/Transform/LandmarkTransform/index.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (joannacirillo) [reply Sources/Common/Transform/LandmarkTransform/index.js:17]: Actually I think it is espacially the rowsToMat3 that is usefull. I created all of them so if in further refactorings they can be used.

File: Sources/Common/Core/Math/index.js (hunk 3 above)
Comment (joannacirillo) [reply Sources/Common/Core/Math/index.js]: Change was made at that place.
